### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [2]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [3]:
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [4]:
total_players = len(purchase_data["SN"].unique())
total_players

576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [5]:
number_items = len(purchase_data["Item ID"].unique())
total_price = sum(purchase_data["Price"].unique())
average_price = total_price/number_items
number_purch = len(purchase_data["Purchase ID"])
total_rev = sum(purchase_data["Price"])
print(number_items)
print(average_price)
print(number_purch)
print(total_rev)
summary_df = pd.DataFrame({
    "Number of Unique Items": [number_items],
    "Average Price": [average_price],
    "Number of Purchases": [number_purch],
    "Total Revenue": [total_rev]
})

#Formatting
summary_df["Average Price"]= summary_df["Average Price"].map("${:,.2f}".format)
summary_df["Total Revenue"]= summary_df["Total Revenue"].map("${:,.2f}".format)
summary_df

183
2.4025683060109286
780
2379.7699999999973


Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $2.40                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [7]:
total_male = len(purchase_data.loc[(purchase_data["Gender"] == "Male")])
total_female = len(purchase_data.loc[(purchase_data["Gender"] == "Female")])
total_other = len(purchase_data.loc[(purchase_data["Gender"] == "Other / Non-Disclosed")])
total_pop = total_male + total_female + total_other

percentage_m = total_male/total_pop * 100
percentage_f = total_female/total_pop * 100
percentage_o = total_other/total_pop * 100

summary = pd.DataFrame({
    "Total Count" : [total_male, total_female, total_other],
    "Percentage of Players" : [percentage_m, percentage_f, percentage_o],
    "Gender" : ["Male", "Female", "Other/ Non-Disclosed"]
})
#formatting
summary = summary.set_index("Gender")
summary["Percentage of Players"] = summary["Percentage of Players"].map("{:.2f}".format)
summary["Percentage of Players"] = summary["Percentage of Players"].astype(str) + '%'
summary

Total Count Percentage of Players
Gender                                                 
Male                          652                83.59%
Female                        113                14.49%
Other/ Non-Disclosed           15                 1.92%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [15]:
female_df = purchase_data.loc[(purchase_data["Gender"] == "Female")]
fnumber_items = len(female_df["Item ID"])
faverage_price = female_df["Price"].mean()
ftotal_rev = sum(female_df["Price"])

male_df = purchase_data.loc[(purchase_data["Gender"] == "Male")]
mnumber_items = len(male_df["Item ID"])
maverage_price = male_df["Price"].mean()
mtotal_rev = sum(male_df["Price"])

o_df = purchase_data.loc[(purchase_data["Gender"] == "Other / Non-Disclosed")]
onumber_items = len(o_df["Item ID"])
oaverage_price = o_df["Price"].mean()
ototal_rev = sum(o_df["Price"])


gendersummary = pd.DataFrame({
    "Purchase Count" : [fnumber_items, mnumber_items, onumber_items],
    "Average Purchase Price" : [faverage_price, maverage_price, oaverage_price],
    "Total Purchase Value" : [ftotal_rev, mtotal_rev, ototal_rev],
    "Gender" : ["Female", "Male", "Other/Non-Disclosed"]
})

#formatting
gendersummary = gendersummary.set_index("Gender")
gendersummary["Average Purchase Price"]= gendersummary["Average Purchase Price"].map("${:,.2f}".format)
gendersummary["Total Purchase Value"]= gendersummary["Total Purchase Value"].map("${:,.2f}".format)
gendersummary


Purchase Count Average Purchase Price  \
Gender                                                       
Female                          113                  $3.20   
Male                            652                  $3.02   
Other/Non-Disclosed              15                  $3.35   

                    Total Purchase Value  
Gender                                    
Female                           $361.94  
Male                           $1,967.64  
Other/Non-Disclosed               $50.19

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [14]:
unique_SN = purchase_data[["SN", "Price", "Age"]]
bins = [0,10,14,19,24,29,34,39,100]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29","30-34","35-39","40+"]
age_bins = unique_SN.groupby([pd.cut(unique_SN["Age"],bins, labels = group_names)])
counts = age_bins.size()
df = pd.DataFrame(counts)
df.columns = ["Total Count"]

perc = df.groupby(["Age"]).agg({"Total Count": "sum"})
perc1 = round(perc.groupby(level=0).apply(lambda x:
                                   100* x / float(df["Total Count"].sum())),2)
df
Age_Demo_table = pd.merge(df, perc1, on="Age")
Age_Demo_table.columns = ["Total Count" , "Percentage of Players"]

#formatting
Age_Demo_table["Percentage of Players"] = Age_Demo_table["Percentage of Players"].map("{:.2f}".format)
Age_Demo_table["Percentage of Players"] = Age_Demo_table["Percentage of Players"].astype(str) + '%'
Age_Demo_table

Total Count Percentage of Players
Age                                     
<10             32                 4.10%
10-14           19                 2.44%
15-19          136                17.44%
20-24          365                46.79%
25-29          101                12.95%
30-34           73                 9.36%
35-39           41                 5.26%
40+             13                 1.67%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [34]:
unique_SN = purchase_data[["SN", "Price", "Age"]]
bins = [0,10,14,19,24,29,34,39,100]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29","30-34","35-39","40+"]
age_bins = unique_SN.groupby([pd.cut(unique_SN["Age"],bins, labels = group_names)])
counts = age_bins.size()
average = age_bins.mean()["Price"]
total = age_bins.sum()["Price"]
df = pd.DataFrame({"Total Count": counts,
                  "Average Price": average,
                  "Average Total per person": total})


#formatting
df["Average Price"] = df["Average Price"].map("${:.2f}".format)
df["Average Total per person"] = df["Average Total per person"].map("${:.2f}".format)
df


Total Count Average Price Average Total per person
Age                                                      
<10             32         $3.40                  $108.96
10-14           19         $2.68                   $50.95
15-19          136         $3.04                  $412.89
20-24          365         $3.05                 $1114.06
25-29          101         $2.90                  $293.00
30-34           73         $2.93                  $214.00
35-39           41         $3.60                  $147.67
40+             13         $2.94                   $38.24

# Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [39]:
preteen_df = purchase_data.loc[(purchase_data["Age"] <= 10)]
preteen_items = len(preteen_df["Item ID"])
paverage_price = preteen_df["Price"].mean()
ptotal_rev = sum(preteen_df["Price"])

teen_df = purchase_data.loc[(purchase_data["Age"] <= 14)]
teen_items = len(teen_df["Item ID"])
taverage_price = teen_df["Price"].mean()
ttotal_rev = sum(teen_df["Price"])

ya_df = purchase_data.loc[(purchase_data["Age"] <= 19)]
ya_items = len(ya_df["Item ID"])
yaaverage_price = ya_df["Price"].mean()
yatotal_rev = sum(ya_df["Price"])

adult_df = purchase_data.loc[(purchase_data["Age"] <= 24)]
adult_items = len(adult_df["Item ID"])
adultaverage_price = adult_df["Price"].mean()
adulttotal_rev = sum(adult_df["Price"])

twenty_df = purchase_data.loc[(purchase_data["Age"] <= 29)]
twenty_items = len(twenty_df["Item ID"])
twenty_price = twenty_df["Price"].mean()
twenty_rev = sum(twenty_df["Price"])

ythir_df = purchase_data.loc[(purchase_data["Age"] <= 34)]
ythir_items = len(ythir_df["Item ID"])
ythir_price = ythir_df["Price"].mean()
ythir_rev = sum(ythir_df["Price"])

thirty_df = purchase_data.loc[(purchase_data["Age"] <= 39)]
thirty_items = len(thirty_df["Item ID"])
thirty_price = thirty_df["Price"].mean()
thirty_rev = sum(thirty_df["Price"])

parents_df = purchase_data.loc[(purchase_data["Age"] >= 40)]
parents_items = len(parents_df["Item ID"])
parents_price = parents_df["Price"].mean()
parents_rev = sum(parents_df["Price"])

agesummary = pd.DataFrame({
    "Purchase Count" : [preteen_items, teen_items, ya_items, adult_items, twenty_items, ythir_items, thirty_items, parents_items],
    "Average Purchase Price" : [paverage_price, taverage_price, yaaverage_price, adultaverage_price, twenty_price, ythir_price, thirty_price, parents_price],
    "Total Purchase Value" : [ptotal_rev, ttotal_rev, yatotal_rev, adulttotal_rev, twenty_rev, ythir_rev, thirty_rev, parents_rev],
    "Age" : ["<10", "10-14", "15-19", "20-24", "25-29","30-34","35-39","40+"]
})

agesummary = agesummary.set_index("Age")
agesummary



Purchase Count  Average Purchase Price  Total Purchase Value
Age                                                                
<10                32                3.405000                108.96
10-14              51                3.135490                159.91
15-19             187                3.063102                572.80
20-24             552                3.055906               1686.86
25-29             653                3.031945               1979.86
30-34             726                3.021846               2193.86
35-39             767                3.052842               2341.53
40+                13                2.941538                 38.24

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [45]:
SN_purchase_count = purchase_data.groupby(["SN"]).count()["Price"].rename("Purchase Count")
SN_average = purchase_data.groupby(["SN"]).mean()["Price"].rename("Average Purchase Price")
SN_total = purchase_data.groupby(["SN"]).sum()["Price"].rename("Total")
sn_df = pd.DataFrame({"Purchase Count": SN_purchase_count,
                     "Average Purchase Price": SN_average,
                     "Total Purchase Value": SN_total})

sn_df["Average Purchase Price"] = sn_df["Average Purchase Price"].map("${:.2f}".format)
sn_df["Total Purchase Value"] = sn_df["Total Purchase Value"].map("${:.2f}".format)
sn_df = sn_df.sort_values(by=["Total Purchase Value"], ascending =False)
sn_df.head()


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                      
Haillyrgue51               3                  $3.17                $9.50
Phistym51                  2                  $4.75                $9.50
Lamil79                    2                  $4.64                $9.29
Aina42                     3                  $3.07                $9.22
Saesrideu94                2                  $4.59                $9.18

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [46]:
purchase_count = purchase_data.groupby(["Item Name", "Item ID"]).count()["Purchase ID"]
items_purchase_value = purchase_data.groupby(["Item Name", "Item ID"]).sum()["Price"]
item_purchase_price = purchase_data.groupby(["Item Name", "Item ID"]).mean()["Price"]
summary = pd.DataFrame({"Purchase Count": purchase_count,
           "Purchase Value": items_purchase_value,
                       "Item Price": item_purchase_price})

#formatting
summary_sort = summary.sort_values("Purchase Count", ascending=False)
summary_sort["Purchase Value"]= summary_sort["Purchase Value"].map("${:,.2f}".format)
summary_sort["Item Price"]= summary_sort["Item Price"].map("${:,.2f}".format)

summary_sort.head()

Purchase Count  \
Item Name                                    Item ID                   
Oathbreaker, Last Hope of the Breaking Storm 178                  12   
Extraction, Quickblade Of Trembling Hands    108                   9   
Nirvana                                      82                    9   
Fiery Glass Crusader                         145                   9   
Pursuit, Cudgel of Necromancy                19                    8   

                                                     Purchase Value Item Price  
Item Name                                    Item ID                            
Oathbreaker, Last Hope of the Breaking Storm 178             $50.76      $4.23  
Extraction, Quickblade Of Trembling Hands    108             $31.77      $3.53  
Nirvana                                      82              $44.10      $4.90  
Fiery Glass Crusader                         145             $41.22      $4.58  
Pursuit, Cudgel of Necromancy                19               $8.16      $1.02

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [58]:
summary_sort = summary.sort_values("Purchase Value", ascending=False)
summary_sort["Purchase Value"]= summary_sort["Purchase Value"].map("${:,.2f}".format)
summary_sort["Item Price"]= summary_sort["Item Price"].map("${:,.2f}".format)
summary_sort.head()

Purchase Count  \
Item Name                                    Item ID                   
Oathbreaker, Last Hope of the Breaking Storm 178                  12   
Nirvana                                      82                    9   
Fiery Glass Crusader                         145                   9   
Final Critic                                 92                    8   
Singed Scalpel                               103                   8   

                                                     Purchase Value Item Price  
Item Name                                    Item ID                            
Oathbreaker, Last Hope of the Breaking Storm 178             $50.76      $4.23  
Nirvana                                      82              $44.10      $4.90  
Fiery Glass Crusader                         145             $41.22      $4.58  
Final Critic                                 92              $39.04      $4.88  
Singed Scalpel                               103             $34.80      $4.35

# Observational Trends

* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  

* The purchases data also reflects the above observation with male gamers spending a total of 1967.64 on game ad ons Which is over five times the amount spent by female gamers $361.94. 

* Even though the peak demographis falls between 20-24 the age demographic that spends the most money is actually 30-34 and 35-40. 

* Oathbreaker, Last Hope of the Breaking Storm is by far the most profitable item purchase. 

-----